bootstrap/src/main/java/jenkins/bootstrap/Bootstrap.java

```diff
@@ -0,0 +1,313 @@
+
+            recordBootAttempt(home);
+
+            coreClassLoader = buildCoreClassLoader();
+            Thread.currentThread().setContextClassLoader(coreClassLoader);
+
+            bootLogics = loadBootLogics(coreClassLoader);
+
+            for (BootLogic b : bootLogics) {
+                b.contextInitialized(event);
+            }
+        } catch (Error | RuntimeException e) {
+            LOGGER.log(SEVERE, "Failed to initialize Jenkins",e);
+            throw e;
+        } catch (Exception e) {
+            LOGGER.log(SEVERE, "Failed to initialize Jenkins",e);
+            throw new Error(e);
+        }
+    }
+
+    public void contextDestroyed(ServletContextEvent event) {
+        for (BootLogic b : bootLogics) {
+            b.contextDestroyed(event);
+        }
+    }
+
+    /**
+     * {@link BootLogic}s that were found, in the order of preference.
+     */
+    public List<BootLogic> getBootLogics() {
+        return Collections.unmodifiableList(bootLogics);
+    }
+
+    /**
+     * ClassLoader that loads Jenkins core.
+     */
+    public ClassLoader getCoreClassLoader() {
+        return coreClassLoader;
+    }
+
+    /**
+     * Location of JENKINS_HOME.
+     */
+    public File getHome() {
+        return home;
+    }
+
+    private ClassLoader buildCoreClassLoader() throws IOException {
+        List<URL> urls = new ArrayList<>();
+        List<DependenciesTxt> overrides = new ArrayList<>();
+
+        DependenciesTxt core = new DependenciesTxt(getClass().getClassLoader().getResourceAsStream("dependencies.txt"));
+
+        File[] plugins = new File(home, "plugins").listFiles();
+        if (plugins!=null) {
+            for (File p : plugins) {
+                String n = p.getName();
+                if (p.isFile() && (n.endsWith(".jpi") || n.endsWith(".hpi"))) {
+                    try {
+                        new Plugin(p).process(core, urls, overrides);
+                    } catch (IOException e) {
+                        LOGGER.log(WARNING, "Skipping "+p, e);
+                    }
+                }
+            }
+        }
+
+        Set<String> jars = context.getResourcePaths("/WEB-INF/jars");
+        if (jars==null || jars.isEmpty()) {
+            throw new IllegalStateException("No WEB-INF/jars");
+        }
+
+        OUTER:
+        for (String jar : jars) {
+            if (!jar.endsWith(".jar"))
+                continue;   // not a jar file
+
+            String justName = jar.substring(jar.lastIndexOf('/')+1);
+
+            Dependency d = core.fromFileName(justName);
+            if (d==null) {
+                // a jar is present in WEB-INF/lib that's not accounted for in dependencies.txt
+                // perhaps somebody manually repackaged a war? let that be in the classpath so as
+                // not to block a desparate attempt by an admin
+                LOGGER.log(INFO, "Allowing unexpected core jar file without override check: "+jar);
```

Comment:
I don't care too strongly one way or the other, but this isn't actionable from users' perspective --- either it's a bug in Jenkins, or s/he intended this to be, so in my mind that makes this INFO.